pkg/reconciler/awss3source/bucket.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright (c) 2021 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awss3source
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/reconciler"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sqs"
+
+	"github.com/triggermesh/aws-event-sources/pkg/apis/sources/v1alpha1"
+	"github.com/triggermesh/aws-event-sources/pkg/reconciler/common/event"
+)
+
+// ensureNotificationsEnabled ensures that event notifications are enabled in
+// the S3 bucket.
+func (r *Reconciler) ensureNotificationsEnabled(ctx context.Context, cli s3iface.S3API, queueARN string) error {
+	src := v1alpha1.SourceFromContext(ctx)
+	typedSrc := src.(*v1alpha1.AWSS3Source)
+
+	status := &typedSrc.Status
+
+	bucketARN := typedSrc.Spec.ARN
+
+	notifCfg, err := getNotificationsConfig(ctx, cli, bucketARN.Resource)
+	switch {
+	case isDenied(err):
+		status.MarkNotSubscribed(v1alpha1.AWSS3ReasonAPIError, "Access denied to S3 API")
+		return controller.NewPermanentError(reconciler.NewEvent(corev1.EventTypeWarning, ReasonFailedSubscribe,
```

Comment:
nit

not wrong but maybe also not right. If a user gets an error contacting the API, would the user fix the issue at AWS and expect that we reconcile automatically, or would they expect to re-create the source after they modify permissions?

Reply:
You're absolutely right. The approach I chose was a trade-off because I don't like error loops :)

There is a refresh configured in the informer for that reason. I have a personal preference for not letting reconcilers enter a loop if the credentials are obviously not authorized, but I get your point and fully agree.

I guess the right approach here would be something in between, like "retry, but only in 30s" or something like that.

The question here is: what do you think will happen more often? Credentials are not valid and need a refresh anyway, or credentials are missing a permission and need to change at AWS? (I don't have the answer)

Reply:
yes, let's resolve this since no one is going to provide the answer.

@JeffNeff  mentioned this recently that he thought that producing error events from sources could be beneficial in some situations ... maybe permanent errors is one.